Refit/Buffers/PooledBufferWriter.Stream.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+
+namespace Refit.Buffers
+{
+    internal sealed partial class PooledBufferWriter
+    {
+        /// <summary>
+        /// An in-memory <see cref="Stream"/> that uses memory buffers rented from a shared pool
+        /// </summary>
+        private sealed class PooledMemoryStream : Stream
```

Comment:
Are these `Stream`s exposed to user? If so they should override the `{Read|Write}Async` as the ones provided by the base `Stream` class are massively inefficient.

Reply:
Hey Ben, thank you for chiming in!
The `Stream` in question is not exposed to users, it's just used internally to create a `StreamContent` instance during the serialization pass, here:

https://github.com/Sergio0694/refit/blob/db28099069707fe7a7136b67c2eadb6a53e7d43a/Refit/SystemTextJsonContentSerializer.cs#L47

That said, you make a good point about the base `WriteAsync` and `ReadAsync` methods being inefficient, and I'm not perfectly sure those wouldn't be used by the consumer of that `StreamContent` instance. Would it make more sense to inherit from `MemoryStream` instead, to leverage the overrides of those methods from that type?

Something like this, with just the added logic to return the buffer:

```csharp
/// <summary>
/// An in-memory <see cref="Stream"/> that uses memory buffers rented from a shared pool
/// </summary>
private sealed class PooledMemoryStream : MemoryStream
{
    private bool isDisposed;

    /// <summary>
    /// Creates a new <see cref="PooledMemoryStream"/> instance.
    /// </summary>
    /// <param name="writer">The input <see cref="PooledBufferWriter"/> instance to read from.</param>
    public PooledMemoryStream(PooledBufferWriter writer)
        : base(writer.buffer, 0, writer.position, false, true)
    {
    }

    /// <inheritdoc/>
    protected override void Dispose(bool disposing)
    {
        base.Dispose(disposing);

        if (isDisposed) return;

        isDisposed = true;

        byte[] buffer = GetBuffer();

        ArrayPool<byte>.Shared.Return(buffer);
    }
}
```

What do you think?

Reply:
No; you'd need to overload them yourself. `MemoryStream` drops back to the inefficient base methods if derived

![image](https://user-images.githubusercontent.com/1142958/77253466-64a6ea00-6c52-11ea-81a6-048a488f44e8.png)

Reply:
Oh, yeah I've seen those checks, I was just not looking at them in this context since `refit` is limited to .NET Standard 2.0, so I wouldn't be able to override those anyway 😟

I've implemented `ReadAsync` and `ReadByte` as you suggested though, in https://github.com/reactiveui/refit/pull/836/commits/1ae2b486d282f3c6cccefe86b77bd6b8afbc825a.

If you'd like me to also implement the new `Read(Span<byte>)` and `ReadAsync(Memory<byte>, CancellationToken)` overloads, I guess I could do that if we also added a `netstandard2.1` target to the library, though I suppose that'd require a major version bump as well? 🤔